server/api/auth/auth.api.spec.js

```diff
@@ -107,49 +35,449 @@ describe('Authentication API', function() {
           valueSet: false
         },
         {
-          message: 'is required',
+          message: 'password or registration OTP is required',
           type: 'json',
-          location: '/password',
-          validator: 'required',
-          valueSet: false
+          location: '',
+          validator: 'auth.credentialsRequired',
+          value: {},
+          valueSet: true
         }
       ]);
     });
 
-    it('should not log in with invalid credentials', async function() {
+    it('should not accept multiple credential types', async function() {
 
-      reqBody.email = 'foo';
-      reqBody.password = '   ';
-      const res = this.test.res = await api.create('/auth', reqBody, { expectedStatus: 422 });
+      const res = this.test.res = await api.create('/auth', {
+        password: 'letmein',
+        registrationOtp: '123456'
+      }, { expectedStatus: 422 });
 
       expectErrors(res, [
         {
-          message: 'must be a valid e-mail address',
+          message: 'is required',
           type: 'json',
           location: '/email',
-          validator: 'email',
-          value: 'foo',
-          valueSet: true
+          validator: 'required',
+          valueSet: false
         },
         {
-          message: 'must not be blank',
+          message: 'password and registration OTP cannot be used at the same time',
           type: 'json',
-          location: '/password',
-          validator: 'notBlank',
-          value: '   ',
+          location: '',
+          validator: 'auth.credentialTypesExclusive',
+          value: {
+            password: 'letmein',
+            registrationOtp: '123456'
+          },
           valueSet: true
         }
       ]);
     });
 
-    /**
-     * Returns an object representing the expected properties of a User, based on the default request body for this test block.
-     *
-     * @param {...Object} changes - Additional expected changes compared to the requets body (merged with Lodash's `extend`).
-     * @returns {Object} An expectations object.
-     */
-    function getExpectedUser(...properties) {
-      return _.extend({}, reqBody, ...properties);
-    }
+    describe('with a password', () => {
+      beforeEach(async () => {
+
+        const password = userFixtures.password();
+        user = await userFixtures.user({
+          createdAt: twoDaysAgo,
+          password: password,
+          updatedAt: twoDaysAgo
+        });
+
+        reqBody = {
+          email: user.get('email'),
+          password: password
+        };
+      });
+
+      it('should log in a user', async function() {
+
+        const res = this.test.res = await api.create('/auth', reqBody);
+
+        const twoWeeksFromNow = moment(now).add(2, 'weeks');
+        await expectApiJwt(res.body.token, user, twoWeeksFromNow);
+
+        await expectUser(res.body.user, getExpectedUser(user, {
+          createdAt: twoDaysAgo,
+          updatedAt: twoDaysAgo
```

Comment:
Actually it is already checked. The `expect*` functions are written so that they always assert something about all properties (whether they are present in the expectations or not).

The `roles` property is asserted to be [empty, unless specified](https://github.com/MediaComem/biopocket-backend/blob/a864b60e782a654489e12df53a71864ff8d8afef/server/spec/expectations/user.js#L88).

The `providerData` property is asserted to [be there or not](https://github.com/MediaComem/biopocket-backend/blob/a864b60e782a654489e12df53a71864ff8d8afef/server/spec/expectations/user.js#L46-L51), and its [value is also checked if it is there](https://github.com/MediaComem/biopocket-backend/blob/a864b60e782a654489e12df53a71864ff8d8afef/server/spec/expectations/user.js#L73-L75). In the case of this test, we would expect it not to be there (it is nullable, and not set by the `local` provider).